File: 2023/day6.py
def race(hold, duration):
    if hold >= duration or hold ==0:
        return 0
    return hold * (duration - hold)

def lower_bound(duration, best):
    for i in range(duration):
        r = race(i, duration)
        if r > best:
            return i
    return False

def upper_bound(duration, best):
    for i in range(duration, 0, -1):
        r = race(i, duration)
        if r > best:
            return i
    return False

def part1():
    t = [35, 69, 68, 87]
    d = [213, 1168, 1086, 1248]
    total = 1
    for i in range(len(t)):
        total *= (upper_bound(t[i], d[i]) - lower_bound(t[i], d[i]) + 1)
    return total

def part2():
    t = 35696887
    d = 213116810861248
    return upper_bound(t, d) - lower_bound(t, d) + 1

print(f'part 1: {part1()} part 2: {part2()}')

File: 2023/go/day6.go
package main

import (
	"fmt"
)

func race(hold, duration int) int {
	if hold >= duration || hold == 0 {
		return 0
	}
	return hold * (duration - hold)
}

func lowerBound(duration, best int) int {
	for i := 0; i < duration; i++ {
		r := race(i, duration)
		if r > best {
			return i
		}
	}
	return -1
}

func upperBound(duration, best int) int {
	for i := duration; i > 0; i-- {
		r := race(i, duration)
		if r > best {
			return i
		}
	}
	return -1
}

func part1() int {
	t := []int{35, 69, 68, 87}
	d := []int{213, 1168, 1086, 1248}
	total := 1
	for i := 0; i < len(t); i++ {
		total *= (upperBound(t[i], d[i]) - lowerBound(t[i], d[i]) + 1)
	}
	return total
}

func part2() int {
	t := 35696887
	d := 213116810861248
	return upperBound(t, d) - lowerBound(t, d) + 1
}

func main() {
	fmt.Printf("part 1: %d part 2: %d\n", part1(), part2())
}
